File: built-in plugins dsp/mb-stereo-imager.cpp
/**
 * MB Stereo Imager
 * Category : effect
 * Type     : eq
 * Version  : 1.0.0
 * Author   : Max Booster
 * Desc     : Multiband stereo width processor for spatial enhancement
 *
 * Part of Max Booster Built-In Plugins DSP
 */

#ifndef MB_STEREO_IMAGER_H
#define MB_STEREO_IMAGER_H

#include <algorithm>
#include <cmath>
#include <cstring>
#include "PluginBase.h"

class MbStereoImager : public PluginBase {
public:
    static constexpr const char* PLUGIN_ID      = "mb-stereo-imager";
    static constexpr const char* PLUGIN_NAME    = "MB Stereo Imager";
    static constexpr const char* PLUGIN_TYPE    = "eq";
    static constexpr const char* PLUGIN_CATEGORY = "effect";
    static constexpr const char* VERSION         = "1.0.0";

    struct Parameters {
    float low_width = 0.5f;  // range [0, 2]
    float mid_width = 1f;  // range [0, 2]
    float high_width = 1.5f;  // range [0, 2]
    float crossover_low = 200f;  // range [50, 500]
    float crossover_high = 5000f;  // range [1000, 15000]
    float mono_below = 80f;  // range [0, 300]
    };

    MbStereoImager() = default;
    ~MbStereoImager() override = default;

    void setSampleRate(double sampleRate) override {
        sampleRate_ = sampleRate;
        reset();
    }

    void reset() override {
        std::memset(buffer_, 0, sizeof(buffer_));
    }

    void process(float* left, float* right, int numSamples, Parameters params) {
        params.low_width = std::clamp(params.low_width, 0f, 2f);
        params.mid_width = std::clamp(params.mid_width, 0f, 2f);
        params.high_width = std::clamp(params.high_width, 0f, 2f);
        params.crossover_low = std::clamp(params.crossover_low, 50f, 500f);
        params.crossover_high = std::clamp(params.crossover_high, 1000f, 15000f);
        params.mono_below = std::clamp(params.mono_below, 0f, 300f);
        for (int i = 0; i < numSamples; ++i) {
            left[i]  = processSample(left[i],  params);
            right[i] = processSample(right[i], params);
        }
    }

private:
    double sampleRate_ = 44100.0;
    float  buffer_[65536] = {};

    inline float processSample(float input, const Parameters& params) {
        // DSP implementation for MB Stereo Imager
        return input;
    }
};

#endif // MB_STEREO_IMAGER_H

File: built-in plugins dsp/mb-stereo-imager.ts
import type { PluginDefinition } from '../server/services/pluginHostService';

const MbStereoImagerPlugin: PluginDefinition = {
    id: 'mb-stereo-imager',
    slug: 'mb-stereo-imager',
    name: 'MB Stereo Imager',
    category: 'effect',
    type: 'eq',
    version: '1.0.0',
    description: 'Multiband stereo width processor for spatial enhancement',
    author: 'Max Booster', grade: 'A',
    parameters: [
      { id: 'low_width', name: 'Low Width', type: 'float', defaultValue: 0.5, minValue: 0, maxValue: 2, automatable: true },
      { id: 'mid_width', name: 'Mid Width', type: 'float', defaultValue: 1, minValue: 0, maxValue: 2, automatable: true },
      { id: 'high_width', name: 'High Width', type: 'float', defaultValue: 1.5, minValue: 0, maxValue: 2, automatable: true },
      { id: 'crossover_low', name: 'Low Crossover', type: 'float', defaultValue: 200, minValue: 50, maxValue: 500, unit: 'Hz', automatable: true },
      { id: 'crossover_high', name: 'High Crossover', type: 'float', defaultValue: 5000, minValue: 1000, maxValue: 15000, unit: 'Hz', automatable: true },
      { id: 'correlation', name: 'Correlation', type: 'bool', defaultValue: true, automatable: false },
      { id: 'mono_below', name: 'Mono Below', type: 'float', defaultValue: 80, minValue: 0, maxValue: 300, unit: 'Hz', automatable: true },
    ],
    defaultPreset: { low_width: 0.5, mid_width: 1, high_width: 1.5, crossover_low: 200, crossover_high: 5000, correlation: true, mono_below: 80 },
  };

export default MbStereoImagerPlugin;
